triplea-game/triplea
# Add new test to be run by Travis: Smoke testing

## Overview
Add a new sub-project for smoke testing. This is end-to-end verification where we avoid
validating logic and we send requests into the system and validate that we get no error
messages.

A travis job is added to run the test. The job is powered by a shell script that coordinates
the steps for the testing:
- start lobby
- start bot
- verify the two have started and are listening on respective ports
- use smoke-test Java client to connect to lobby
- use smoke-test Java client to connect to bot
- verify logs of bot and lobby do not contain any SEVERE logging.

If the bot fails to connect to lobby, it will continue running but will log a SEVERE error.
Thus, we check logs to detect that condition.

If either lobby or bot fail to startup completely, we'll get a fail fast error using 'nc' (netcat)
which checks if the lobby or bot ports are open.

On exit, bot and lobby are stopped and logs are printed to help with debug. We will always print logs so that on successful builds we have a point of comparison, particularly to help compare against failure builds.

Note:
- The smoke test is not integrated with gradle check command as it depends on the current system state and closes current lobby and bot. If there is any manual testing then this could quickly get in the way.



## Functional Changes

[ ] New map or map update
[ ] New Feature
[ ] Feature update or enhancement
[ ] Feature Removal
[ ] Code Cleanup or refactor
[x] Configuration Change
[ ] Bug fix





## Testing


[ ] Covered by existing automated tests
[ ] Covered by newly added automated tests
[x] Manually tested
[ ] No testing done

### Manual Testing

- Ran locally, verified various failure permutations are spotted by the test. For example, no lobby running, lobby failing to run due to DB down, no bot running
- An example of the job output can be found at: https://travis-ci.org/DanVanAtta/triplea/jobs/580343505  (or from this branch PR build)
